Rewrite all material including README and examples

robertmcnees opened the pull request:
This PR is a near complete rewrite of the existing guide. 

The existing guide was dated in several ways, most notably in a reference to `@EnableDiscoveryClient` (identified in issue #25) and `boostrap.properties` (identified in issue #28). Both of these methods are not available or recommended with the latest version of Spring Cloud. Rather than simply updating the existing applications to work, I took the opportunity to rethink the examples used.

A couple of key differences in this PR to what I replaced:

- Removed the `complete` and `initial` folders: I felt that having multiple applications in a single guide felt very cluttered when they were repeated in each folder. The `complete` and `initial` format works OK for simple apps, but when multiple apps are required it is my opinion that complexity increases beyond the value. If the users want to start from scratch, they can use the provided Spring Initializr links.
- Added a third application: I felt that the existing example didn't go the final step and help the user understand how service discovery could be used in a real application. To demonstrate this, I created two services that are both registered with Eureka and communicate to one another.
- I used a single build tool: For the same reason that I removed `complete` and `initial` folders, I decided against including both gradle and maven dependencies. There are several build commands specified in the `README.adoc` and I wanted to avoid having to list both gradle and maven versions.
- Added a CI Build in GitHub Actions, similar to other spring-guides.

robertmcnees commented:
Given the significant changes, I'm hoping to have a review from @OlgaMaciaszek.

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
+This guide needs three applications.
+The first application (the server application) needs
 only the Eureka Server dependency.
+The second and third applications (the client applications) need Eureka Discovery Client and Spring Web dependencies.
```
Add "the" before "Eureka"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
+- {clientb-link}[Service B^]
 
-You can use this {server-link}[pre-initialized project] (for the service application) or this {client-link}[pre-initialized project] (for the client application) and click Generate to download a ZIP file. This project is configured to fit the examples in this tutorial.
+NOTE: Due to the number of services in this guide, only the solutions are provided in the GitHub repository. If you'd like to start from scratch, please use generate the blank projects using the links above or using https://start.spring.io[Spring Initializr] as defined below.
```
Replace "if you'd like to" with "To"

Remove "please use"

Add a comma before "as defined below"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
 This service pulls in all the dependencies you need for an application and does most of the setup for you.
 . Choose either Gradle or Maven and the language you want to use. This guide assumes that you chose Java.
-. Click *Dependencies* and select *Eureka Server* for the service application and *Eureka Server* and *Eureka Discovery Client* for the client application.
+. Click *Dependencies* and select *Eureka Server* for the server application. Select *Eureka Discovery Client* and *Spring Web* for the client applications.
```
There should be two sets of instructions here, one for the server and one for the clients.

robertmcnees replied:
I made the change. No objections having 2 lists would increase the clarity. I originally opted for a single list, as 90% of the instructions are the same and I don't think many users will follow the start.spring.io path anyhow.
I can see an argument either way, so I made the change. If it looks too bloated we can always revert.

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
-nodes to which the registry can connect. In a production environment, you will want more
-than one instance of the registry. For our simple purposes, however, it suffices to
-disable the relevant logging.
+In a production environment, you will likely want more than one instance of the registry.
```
Remove "will"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
 ----
 ====
 
+You can now run the Eureka Server with the command `./gradlew bootRun`.
```
Change to:

You can now start the Eureka server by running `./gradlew bootRun`.

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
-The following listing (from
-`eureka-client/src/main/java/example/serviceregistrationanddiscoveryclient/ServiceRegistrationAndDiscoveryClientApplication.java`)
-shows the client application:
+Now that you have started a service registry, you can stand up clients that interact with the registry. Our client applications, ServiceA and ServiceB, will automatically register with the Eureka server because we have `spring-cloud-starter-netflix-eureka-client` on the classpath.
```
Remove "will"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
-`http://localhost:8080/service-instances/a-bootiful-client`. See the
-https://spring.io/guides/gs/rest-service/[Building a RESTful Web Service]  guide to learn
-more about building REST services with Spring MVC and Spring Boot.
+At this point you should be able to run all three applications.
```
Add a comma after "point"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
-https://spring.io/guides/gs/rest-service/[Building a RESTful Web Service]  guide to learn
-more about building REST services with Spring MVC and Spring Boot.
+At this point you should be able to run all three applications.
+You can use the IDE or execute the command `./gradlew bootRun` from each application folder.
```
Move command to after the command.

So replace "command `./gradlew bootRun`" with "`./gradlew bootRun` command"

The other way to do it is "command, `./gradlew bootRun`," (that is, wrapping the command in commas). This has to do with English's ordering of nouns. If you shift from the usual order, you have to make it be parenthetical, so you need commas. I run into this constantly, because people speak in this order (ignoring the commas when speaking), but it's incorrect to write in this order.

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
+You can use the IDE or execute the command `./gradlew bootRun` from each application folder.
 
-== Test the Application
+When the applications are running, you can view the http://localhost:8761/[Eureka dashboard^].
```
Change "are running" to "run"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
 
-Test the end-to-end result by starting the `eureka-service` first and then, once that has
-loaded, starting the `eureka-client`.
+== Service A Endpoint
```
Change "A" to "An"

robertmcnees replied:
The Endpoint is called "Service A". I'm assuming the header is OK in general, as "Service B Endpoint" got no comments. I'll leave this as it is.

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
 
-To run the Eureka service with Maven, run the following command in a terminal window (in
-the `/complete` directory):
+Create a new class `com/example/servicea/controller/ServiceARestController.java` in the `servicea` project that exposes an endpoint you will use to test the application.
```
Add "called" after "class"

Change "that exposes" to "to expose"

Change "will" to "can"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
-the `/complete` directory):
+== Service B Endpoint
+
+Create a new class `com/example/serviceb/controller/ServiceBRestController.java` in the `serviceb` project that exposes an endpoint that calls `servicea`.
```
Add "called" after "class"

Change "that exposes" to "to expose"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
-To run the Eureka service with Gradle, run the following command in a terminal window (in
-the `/complete` directory):
+This class uses a `DiscoveryClient` to find the `ServiceInstance` of `servicea` based only on the application name.
+This guide only has one instance of `servicea`, so we can look only at the first instance.
```
Change "only has" to "has only"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
-the `/complete` directory):
+This class uses a `DiscoveryClient` to find the `ServiceInstance` of `servicea` based only on the application name.
+This guide only has one instance of `servicea`, so we can look only at the first instance.
+This is seen in the line:
```
Change "line" to "following line"

Buzzardo commented on README.adoc:
````diff
@@ -2,188 +2,184 @@
+ServiceInstance serviceInstance = discoveryClient.getInstances("servicea").get(0);
+```
+
+Once you have the `ServiceInstance` that refers to the location of `servicea`, you can now use the information in Spring Framework's https://docs.spring.io/spring-framework/reference/integration/rest-clients.html#rest-restclient[Rest Client].
````
Change "have the" to "have a"

Buzzardo commented on README.adoc:
````diff
@@ -2,188 +2,184 @@
+
+```
+
+You can test all three applications by executing the command:
````
Change "executing" to "running"

Add "following" before "command"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
-see an empty `<List>` element, wait a bit and refresh the page.
+== Test the Application
+
+When you have implemented the code above, or you are using the solution repository, here is an outline of the steps in this guide:
```
Change to:

This guide walked through the following steps (which you could do by implementing the code shown throughout this guide or by using the code in the solution repository):

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
+
+When you have implemented the code above, or you are using the solution repository, here is an outline of the steps in this guide:
+
+- Run `eureka-server` with the command `./gradlew bootRun` from the `eureka-server` folder
```
Change "command `./gradlew bootRun`" to "`./gradlew bootRun` command"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
+When you have implemented the code above, or you are using the solution repository, here is an outline of the steps in this guide:
+
+- Run `eureka-server` with the command `./gradlew bootRun` from the `eureka-server` folder
+- Run `servicea` with the command `./gradlew bootRun` from the `servicea` folder
```
Change "command `./gradlew bootRun`" to "`./gradlew bootRun` command"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
+
+- Run `eureka-server` with the command `./gradlew bootRun` from the `eureka-server` folder
+- Run `servicea` with the command `./gradlew bootRun` from the `servicea` folder
+- Run `serviceb` with the command `./gradlew bootRun` from the `serviceb` folder
```
Change "command `./gradlew bootRun`" to "`./gradlew bootRun` command"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
+- Run `eureka-server` with the command `./gradlew bootRun` from the `eureka-server` folder
+- Run `servicea` with the command `./gradlew bootRun` from the `servicea` folder
+- Run `serviceb` with the command `./gradlew bootRun` from the `serviceb` folder
+- Observe `servicea` and `serviceb` are registered by viewing the Eureka Dahsboard at http://localhost:8761/
```
Change "Observe" to "Observe that"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
+- Run `servicea` with the command `./gradlew bootRun` from the `servicea` folder
+- Run `serviceb` with the command `./gradlew bootRun` from the `serviceb` folder
+- Observe `servicea` and `serviceb` are registered by viewing the Eureka Dahsboard at http://localhost:8761/
+- Run the command `curl http://localhost:8082/serviceB/helloEureka` to test all three applications are working properly
```
Change "command `curl http://localhost:8082/serviceB/helloEureka`" to "`curl http://localhost:8082/serviceB/helloEureka` command"

Buzzardo commented on README.adoc:
```diff
@@ -2,188 +2,184 @@
+- Run `serviceb` with the command `./gradlew bootRun` from the `serviceb` folder
+- Observe `servicea` and `serviceb` are registered by viewing the Eureka Dahsboard at http://localhost:8761/
+- Run the command `curl http://localhost:8082/serviceB/helloEureka` to test all three applications are working properly
+- Observe the output `Hello world from Service A!`
```
Add a comma after "output"

codsaf commented:
### **It is so a nice work!**

codsaf commented:
Can you tell me how the servicea and serviceb know the eureka server's ip address and port?

OlgaMaciaszek commented on README.adoc:
```diff
@@ -2,188 +2,194 @@
+[source,properties]
 ----
-include::complete/eureka-client/src/main/java/com/example/serviceregistrationanddiscoveryclient/ServiceRegistrationAndDiscoveryClientApplication.java[]
+include::servicea/src/main/resources/application.properties[]
```
Since you've rewritten the whole thing, it might also make sense to use `.yml` instead of properties in one or both services.

OlgaMaciaszek commented on README.adoc:
```diff
@@ -2,188 +2,194 @@
 
-To run the Eureka service with Gradle, run the following command in a terminal window (in
-the `/complete` directory):
+This class uses a `DiscoveryClient` to find the `ServiceInstance` of `servicea` based only on the application name.
```
We usually refer to it as `serviceId` ratherthan application name while discussing service discovery.

OlgaMaciaszek commented on README.adoc:
````diff
@@ -2,188 +2,194 @@
+```
+
+... = restClient.get()
+				.uri(serviceInstance.getUri() + "/helloWorld")...
````
There seems to be some formatting issue here.

OlgaMaciaszek commented on serviceb/src/main/java/com/example/serviceb/controller/ServiceBRestController.java:
```diff
@@ -0,0 +1,29 @@
+@RestController
+public class ServiceBRestController {
+
+	private DiscoveryClient discoveryClient;
```
both fields could be `private final`

OlgaMaciaszek commented on serviceb/src/main/java/com/example/serviceb/controller/ServiceBRestController.java:
```diff
@@ -0,0 +1,29 @@
+
+	@GetMapping("helloEureka")
+	public String helloWorld() {
+		ServiceInstance serviceInstance = discoveryClient.getInstances("servicea").get(0);
```
Might make sense to verify if we have found at least one instance before running the call.

robertmcnees replied:
I think that's good defensive coding and something users should do in any example that isn't 'hello world' complexity. I'm not sure if it belongs in the guide.

If I were to expand on the code, I may do:
```
List<ServiceInstance> serviceInstances = discoveryClient.getInstances("servicea");
if(serviceInstances == null || serviceInstances.size() == 0) {
  return "no service instances found";
}
ServiceInstance serviceInstance = serviceInstances.get(0);
String serviceAResponse = ...
```

It feels like there are a lot of ways to do the check. I try for simple and verbose in the guides, so I'd turn 1 line of code into 5 to do the check. Before this edit, I haven't defined an error or fallback value if we have a problem. I could have the method throw an exception, but that adds code to the guide that isn't related to the topic and I don't explain in the README.

In my opinion I'd leave out the defensive check. A lot of our guides opt for the simplest route and leave out defensive coding practices. Current behavior if `servicea` is not available is:
```
 curl http://localhost:8082/helloEureka
{"timestamp":"2024-06-19T11:20:25.727+00:00","status":500,"error":"Internal Server Error","path":"/helloEureka"}
```
And the logs can get you there pretty quick:
```
2024-06-19T07:20:25.714-04:00 ERROR 1649 --- [serviceb] [nio-8082-exec-1] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: java.lang.IndexOutOfBoundsException: Index: 0] with root cause

java.lang.IndexOutOfBoundsException: Index: 0
        at java.base/java.util.Collections$EmptyList.get(Collections.java:4586) ~[na:na]
        at com.example.serviceb.controller.ServiceBRestController.helloWorld(ServiceBRestController.java:22) ~[main/:na]
```

I'll defer to your opinion as the final word. Do you think making the check is worth the extra code? Do you have a preference on how to make the check?

OlgaMaciaszek replied:
Ok, we can leave it as is.

robertmcnees commented:
@OlgaMaciaszek I made all of the changes you requested, with the exception of the defensive coding check I'd like to get your thoughts on.

For the Gradle/Maven discussion, that was a miss on my part. I usually try to include build files in the code for both Gradle and Maven. I like to only have a single set of commands reference in the `README`, and I should default to the preference of the underlying project. In this case, I should be using Maven since that is what Spring Cloud Netflix uses.

I added the Maven wrapper and `pom.xml` to all 3 projects. I also changed the `README` so that we use Maven `./mvnw spring-boot:run` instead of Gradle `./gradlew bootRun`.

OlgaMaciaszek commented:
@robertmcnees there still seem to be some gradle files present in the project. Is this on purpose?

robertmcnees commented:
@OlgaMaciaszek Yes, the inclusion of both gradle and maven is on purpose. I added both build systems here for consistency to our other guides which also include both Maven and Gradle build systems. The README should feature Maven commands, according to your preference.

I believe all of our guides originated with advice from this [wiki](https://github.com/spring-guides/getting-started-guides/wiki) about cloning [this project](https://github.com/spring-guides/getting-started-guides). Several years ago the decision was made to have both build systems available in the guides.

If you feel strongly we should remove Gradle completely from this guide I can go that route. Personally, I'd prefer consistency if it doesn't add complexity for the learner. But I did break tradition here by removing the `complete` and `initial` folders because I felt that repeating all three applications in two places would be too confusing. If there is a similar argument to removing Gradle, then no problem for me to remove it.

From a sustainability perspective, I have GitHub Actions and Dependabot configured so I can manage both build systems simultaneously.

What do you think?

OlgaMaciaszek commented:
Yes, it's good. May be helpful for the users to have both build systems.
